### ISSUE TITLE
False positive LessSpecificReturnStatement/MoreSpecificReturnType when using non-empty-array

### ISSUE THREAD

**Issue by WalentinG** [vimeo/psalm#6458]
```php
<?php

interface I
{
    /** @psalm-return non-empty-array<string, string> */
	public static function methodName(bool $b): array;
}

class C implements I 
{
    public static function methodName(bool $b): array
    {
    	return $b ? ['true' => 'value'] : ['false' => 'value'];
    }
}
```

https://psalm.dev/r/0d1f5fbad4

**Comment by orklah** [comment]
That's a tough one! Neither key is guaranteed to exist but at least a key is guaranteed to exist. Psalm's type system is not subtle enough to see that I think

**Comment by weirdan** [comment]
Also applies to `if/else` (I expected it wouldn't): https://psalm.dev/r/b14298a12f

**Comment by orklah** [comment]
If we wanted to handle that, I think it would need a flag on TKeyedArray like nonEmpty that we could set when we combine two non-empty Keyed Arrays into another one.

However, that would mean maintaining this flag up to date no matter what. For example
`$b ? ['true' => 'value'] : ['false' => 'value'];
unset($b['false']);`
would reset the nonEmpty flag

**Comment by orklah** [comment]
interestingly, PHPStan does this with an union of two Keyed arrays: https://phpstan.org/r/190e6feb-745c-4350-a19e-e421b9bcee2d

It's technically correct and it guarantees the existance of at least a key. We should do something similar if we want to solve this
